hudi-common/src/main/java/org/apache/hudi/metadata/BaseFileRecordParsingUtils.java

```diff
@@ -0,0 +1,186 @@
+                                                              String partition,
+                                                              String latestFileName,
+                                                              String prevFileName,
+                                                              HoodieStorage storage,
+                                                              Set<RecordStatus> recordStatusesOfInterest) throws IOException {
+    Set<String> recordKeysFromLatestBaseFile = getRecordKeysFromBaseFile(storage, basePath, partition, latestFileName);
+    if (prevFileName == null) {
+      if (recordStatusesOfInterest.contains(RecordStatus.INSERT)) {
+        return Collections.singletonMap(RecordStatus.INSERT, new ArrayList<>(recordKeysFromLatestBaseFile));
+      } else {
+        // if this is a new base file for a new file group, everything is an insert.
+        return Collections.emptyMap();
+      }
+    } else {
+      // read from previous base file and find difference to also generate delete records.
+      // we will return updates and deletes from this code block
+      Set<String> recordKeysFromPreviousBaseFile = getRecordKeysFromBaseFile(storage, basePath, partition, prevFileName);
+      Map<RecordStatus, List<String>> toReturn = new HashMap<>(recordStatusesOfInterest.size());
+      if (recordStatusesOfInterest.contains(RecordStatus.DELETE)) {
+        toReturn.put(RecordStatus.DELETE, recordKeysFromPreviousBaseFile.stream()
+            .filter(recordKey -> {
+              // deleted record
+              return !recordKeysFromLatestBaseFile.contains(recordKey);
+            }).collect(toList()));
+      }
+
+      List<String> updates = new ArrayList<>();
+      List<String> inserts = new ArrayList<>();
+      // do one pass and collect list of inserts and updates.
+      recordKeysFromLatestBaseFile.stream().forEach(recordKey -> {
+        if (recordKeysFromPreviousBaseFile.contains(recordKey)) {
+          updates.add(recordKey);
+        } else {
+          inserts.add(recordKey);
+        }
+      });
+      if (recordStatusesOfInterest.contains(RecordStatus.UPDATE)) {
+        toReturn.put(RecordStatus.UPDATE, updates);
+      }
+      if (recordStatusesOfInterest.contains(RecordStatus.INSERT)) {
+        toReturn.put(RecordStatus.INSERT, inserts);
+      }
+      return toReturn;
+    }
+  }
+
+  private static Set<String> getRecordKeysFromBaseFile(HoodieStorage storage, String basePath, String partition, String fileName) throws IOException {
+    StoragePath dataFilePath = new StoragePath(basePath, StringUtils.isNullOrEmpty(partition) ? fileName : (partition + Path.SEPARATOR) + fileName);
+    FileFormatUtils fileFormatUtils = HoodieIOFactory.getIOFactory(storage).getFileFormatUtils(HoodieFileFormat.PARQUET);
+    return fileFormatUtils.readRowKeys(storage, dataFilePath);
+  }
+
+  public enum RecordStatus {
```

Comment:
Kind of personal, just use regular flag in the methods: `auditInsert`? `auditIUpdate` and `auditIDelete`, I don't think the enum and map#contains check is necessary or helps much.

Reply:
Anything more than 2, I would prefer to add enum as a good programming practice.